File: src/main/java/dev/goerner/geozen/model/jackson/serializer/AbstractGeometrySerializer.java
package dev.goerner.geozen.model.jackson.serializer;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.JsonSerializer;
import dev.goerner.geozen.model.Geometry;
import dev.goerner.geozen.model.Position;

import java.io.IOException;

public abstract class AbstractGeometrySerializer<T extends Geometry> extends JsonSerializer<T> {

	protected void writePosition(Position position, JsonGenerator gen) throws IOException {
		gen.writeStartArray();
		gen.writeNumber(position.getLongitude());
		gen.writeNumber(position.getLatitude());
		if (position.getAltitude() != 0) {
			gen.writeNumber(position.getAltitude());
		}
		gen.writeEndArray();
	}
}

File: src/main/java/dev/goerner/geozen/jackson/serializer/AbstractGeometrySerializer.kt
package dev.goerner.geozen.jackson.serializer

import dev.goerner.geozen.model.Geometry
import dev.goerner.geozen.model.Position
import tools.jackson.core.JsonGenerator
import tools.jackson.databind.ValueSerializer

abstract class AbstractGeometrySerializer<T : Geometry> : ValueSerializer<T>() {

    protected fun writePosition(position: Position, gen: JsonGenerator) {
        gen.writeStartArray()
        gen.writeNumber(position.longitude)
        gen.writeNumber(position.latitude)
        if (position.altitude != 0.0) {
            gen.writeNumber(position.altitude)
        }
        gen.writeEndArray()
    }
}
